Rename InputList to DataContainer

pmrv opened the pull request:
As we discussed in the software dev meeting, the name `InputList` is getting a bit outdated, so I changed it.  I'll leave this as a draft until @muh-hassani and I finish the work on reading/writing text files from the container.

pmrv committed:
Rename InputList to DataContainer

max-hassani commented:
> 
> 
> As we discussed in the software dev meeting, the name `InputList` is getting a bit outdated, so I changed it. I'll leave this as a draft until @muh-hassani and I finish the work on reading/writing text files from the container.

The Last two days, I was busy with presentations I had to make for a PMD project meeting. If you give me an hour, I will commit and push my changes to this branch.

pmrv commented:
> > As we discussed in the software dev meeting, the name `InputList` is getting a bit outdated, so I changed it. I'll leave this as a draft until @muh-hassani and I finish the work on reading/writing text files from the container.
> 
> The Last two days, I was busy with presentations I had to make for a PMD project meeting. If you give me an hour, I will commit and push my changes to this branch.

Don't worry, no hurry on my side.

max-hassani committed:
update datacontainer.py with read() and parse_yml()

max-hassani committed:
update datacontainer.py adding Docstring

max-hassani commented:
I will also add a xml parser which returns a dictionary as the yaml parser does.

max-hassani committed:
updating datacontainer.py and setup.py; adding xml parser

max-hassani committed:
updating datacontainer.py

max-hassani committed:
updating .ci_support/environment.yml

jan-janssen commented on .ci_support/environment.yml:
```diff
@@ -19,3 +19,4 @@ dependencies:
   - six >=1.14.0
   - sqlalchemy >=1.3.14
   - tqdm >=4.35.0
+  - xmltodict >=0.11.0
```
Can you try using `defusedxml` because we have made good experience in parsing the VASP xml file. In addition we should add `pyyaml` if we want to support `yaml` import.

max-hassani replied:
I understand that `defusedxml` introduces security over `xml` library, but casting the content of the parser to a dictionary is not straight forward in both cases. On the other hand `xmltodict`, [while it imports from `defusedexpat`](https://github.com/martinblech/xmltodict/blob/ae19c452ca000bf243bfc16274c060bf3bf7cf51/xmltodict.py#L5), make this casting really easy.

max-hassani replied:
Nevertheless, using `defusedxml`, the following `dictify()` function can also turn the content of the parsed xml file to a dictionary:
```
from defusedxml import ElementTree
from copy import copy

def dictify(r,root=True):
    if root:
        return {r.tag : dictify(r, False)}
    d=copy(r.attrib)
    if str(r.text).find('\n') == -1:
        d[r.tag]=r.text
    for x in r.findall("./*"):
        if len(x) >= 1:
            if x.tag not in d:
                d[x.tag]=[]
            d[x.tag].append(dictify(x,False))
        else:
            d[x.tag] = x.text
    return d

root = ElementTree.parse('sample.xml').getroot()
final_dict = dictify(root)
```

jan-janssen replied:
@sudarsan-surendralal Can you comment on the choice of `defusedxml` vs `xmltodict` ? Maybe it is only related to the broken xml format in VASP.

sudarsan-surendralal replied:
`defusedxml` is more secure for parsing xml files and was suggested by codeacy which complained when we used `xmltodict` earlier. We used it to parse `vasprun.xml` files for every single VASP calculation and it works pretty hassle-free. But yes, you need to tweak it a little bit to get the contents into a dictionary form

max-hassani replied:
@sudarsan-surendralal, thanks for your response. I guess in the `xml` files for vasp, you have expected keys, and you look for some values. Am I right? 
But in our case, we need to cast the parsed xml into a dictionary or some sort of data containers to be saved as Datacontainer object.   
Another point, here, we have used `xmltodict` and so far Codacy has not complained about it. It might be that `xmltodict` has been updated.

max-hassani replied:
> `vasprun`

@sudarsan-surendralal, I could not find the code where you parsed `vasprun.xml` under `pyiron_atomistic/sphinx/`. Can you send me a link where `defusedxml` is used?

sudarsan-surendralal replied:
> 
> 
> @sudarsan-surendralal, thanks for your response. I guess in the `xml` files for vasp, you have expected keys, and you look for some values. Am I right?

Yeah, that's true.

> But in our case, we need to cast the parsed xml into a dictionary or some sort of data containers to be saved as Datacontainer object.
> Another point, here, we have used `xmltodict` and so far Codacy has not complained about it. It might be that `xmltodict` has been updated.

Oh ok I was not aware of this. In that case maybe `xmltodict` is more appropriate for you. 

> @sudarsan-surendralal, I could not find the code where you parsed `vasprun.xml` under `pyiron_atomistic/sphinx/`. Can you send me a link where `defusedxml` is used?

It's in this file in the vasp directory https://github.com/pyiron/pyiron_atomistics/blob/master/pyiron_atomistics/vasp/vasprun.py

pmrv committed:
Allow backwards-compatible import of the InputList

pmrv committed:
Clean up read method

pmrv committed:
Fix wordings

jan-janssen commented on pyiron_base/generic/inputlist.py:
```diff
@@ -1,656 +1,5 @@
-            :class:`list`: list of all keys to elements of :class:`InputList`.
-        """
-        return list(self.groups())
+from .datacontainer import DataContainer as InputList
```
I would derive a class named `InputList` from the `DataContainer`, then once this class is used we can print a warning.

jan-janssen commented:
I am fine with using `xmltodict` we should just make sure that we merge with the master branch before to guarantee the dependencies are strict to work with dependabot.

jan-janssen committed:
Merge branch 'master' into data

jan-janssen commented:
@pmrv Do you think we can merge this on Friday? I guess it makes sense to release a new `pyiron_base` version and then update all classes which depend on it.

pmrv commented:
> @pmrv Do you think we can merge this on Friday? I guess it makes sense to release a new `pyiron_base` version and then update all classes which depend on it.

I guess so, biggest thing would be to update the outstanding PR, but only the atomicrex is the hard one.  @muh-hassani What's the status on the xml writing?

max-hassani commented:
> 
> 
> > @pmrv Do you think we can merge this on Friday? I guess it makes sense to release a new `pyiron_base` version and then update all classes which depend on it.
> 
> I guess so, biggest thing would be to update the outstanding PR, but only the atomicrex is the hard one. @muh-hassani What's the status on the xml writing?

I am working on it. write to yml was easy, but the xml one is a bit tricky. I will push it in less than an hour.

pmrv commented:
> > > @pmrv Do you think we can merge this on Friday? I guess it makes sense to release a new `pyiron_base` version and then update all classes which depend on it.
> > 
> > 
> > I guess so, biggest thing would be to update the outstanding PR, but only the atomicrex is the hard one. @muh-hassani What's the status on the xml writing?
> 
> I am working on it. I will push it in less than an hour.

No worries, we will want to discuss on Friday how to update the other repositories & PRs anyway.

max-hassani committed:
adding write yml/xml

max-hassani committed:
adding dicttoxml to setup.py

max-hassani commented:
@pmrv, I added the two functions for writing to yml/xml file. One issue that I have is that datacontainer can have other format than a dictionary, but these two function requires that the datacontainer has some keys and some values in order to be stored in a yml/xml file. I think via `isinstance`, I can not check this, since it is an instance of datacontainer. But maybe checking whether it has keys as attribute could verify that the datacontainer has some keys and values.

max-hassani committed:
updating datacontainer.py

max-hassani committed:
updating .ci_support/environment.yml

pmrv commented:
> @pmrv, I added the two functions for writing to yml/xml file. One issue that I have is that datacontainer can have other format than a dictionary, but these two function requires that the datacontainer has some keys and some values in order to be stored in a yml/xml file. I think via `isinstance`, I can not check this, since it is an instance of datacontainer. But maybe checking whether it has keys as attribute could verify that the datacontainer has some keys and values.

You can check whether the `DataContainer` has non-integer keys with `.has_keys()`, but you should also be able to just convert integer keys to strings and then use them in the xml.

jan-janssen commented on pyiron_base/generic/datacontainer.py:
```diff
@@ -0,0 +1,730 @@
+        xml_data = dicttoxml(self, attr_type=attr_flag)
+        xmlfile = open(file_name, "w")
+        xmlfile.write(parseString(xml_data).toprettyxml())
+        xmlfile.close()
```
Does this work with `with`?
```suggestion
        with open(file_name, "w") as xmlfile:
            xmlfile.write(parseString(xml_data).toprettyxml())
```

pmrv commented:
There should also be some tests with an simple example list and yaml/xml file to check that writing/parsing is lossless for all combinations of having only integer keys, some string keys and all string keys.

max-hassani committed:
Update pyiron_base/generic/datacontainer.py

Co-authored-by: Jan Janssen <jan-janssen@users.noreply.github.com>

max-hassani committed:
updating datacontainer.py

jan-janssen commented on pyiron_base/generic/datacontainer.py:
```diff
@@ -56,6 +57,7 @@ def _parse_yaml(file_name):
             return {}
 
+
 def _parse_xml(self, file_name, wrap=False):
```
```suggestion
def _parse_xml(file_name, wrap=False):
```
Is there are particular reason why the parse functions are not static functions of the class.

pmrv replied:
I didn't want to clutter the interface too much.  If you don't want the function outside the class, I can define them inside the `read()`. 

The `self` slipped through when copying, will fix.

max-hassani committed:
updating datacontainer.py

pmrv committed:
Drop self argument

Co-authored-by: Jan Janssen <jan-janssen@users.noreply.github.com>

max-hassani committed:
Merge branch 'data' of github.com:pyiron/pyiron_base into data

max-hassani committed:
updating datacontainer.py; fixing the issue with xml read/write

max-hassani committed:
updating test_inputlist.py; adding the test for read/write to/from files

max-hassani committed:
updating test_inputlist.py

max-hassani committed:
updating datacontainer.py, removing extra functions

jan-janssen committed:
Merge branch 'master' into data

jan-janssen committed:
Update test_inputlist.py

jan-janssen committed:
Merge branch 'master' into data

max-hassani committed:
update datacontainer.py; fixing issue with lists

max-hassani committed:
updating unittest test_inputlist.py

max-hassani committed:
fixing test_inputlist.py

max-hassani commented:
@pmrv, by adding another function, I fixed the issue with the "item" key in the case of a nested list. If you find another solution, feel free to remove it.

max-hassani committed:
fixing codacy style issues

pmrv committed:
Refactor DataContainer I/O

pmrv commented:
> @pmrv, by adding another function, I fixed the issue with the "item" key in the case of a nested list. If you find another solution, feel free to remove it.

Thanks for fixing this!  I'm ok leaving this as is, I refactored the functions into their own module and will update the tests tomorrow.

pmrv committed:
Rename fileparsers module

jan-janssen commented:
I fixed the tests, so from my perspective it is now ready to merge. Once this is merged then we should release `pyiron_base` version `0.2.0`.

pmrv commented:
> I fixed the tests, so from my perspective it is now ready to merge. Once this is merged then we should release `pyiron_base` version `0.2.0`.

I did some more refactors, will merge after I pushed them.

pmrv committed:
Finish fileio refactor

simplify unit tests, add tests for fileio specifically
refactor reading/writing of files into FileAdapter class
fix XML parsing when keys are numerical strings

pmrv committed:
Touch up docstrings

pmrv committed:
Rename datacontainer tests

pmrv committed:
Add deprecation warning to InputList

pmrv committed:
Add test for DeprecationWarning

pmrv committed:
Remove unused variables

pmrv commented:
I had to force push, because the fixed didn't line up anymore of the refactored `fileio`.  I also found one more bug, when writing dicts to xml with "numeric key" like `"0"` and added a test case for it.

From my side this is ready to go.

jan-janssen commented:
> From my side this is ready to go.

The I would suggest we merge it now and release a new `pyiron_base` version to push the migration for `pyiron_atomistics` and `pyiron_contrib`.

pmrv committed:
Fix imports

pmrv commented:
> > From my side this is ready to go.
> 
> The I would suggest we merge it now and release a new `pyiron_base` version to push the migration for `pyiron_atomistics` and `pyiron_contrib`.

I'll update and merge #70 first and then bump the tag.
